jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/RedshiftUtils.kt

```diff
@@ -3,19 +3,39 @@
 
 package software.aws.toolkits.jetbrains.services.redshift
 
+import com.intellij.database.autoconfig.DataSourceRegistry
 import com.intellij.openapi.project.Project
 import software.amazon.awssdk.services.redshift.model.Cluster
 import software.aws.toolkits.core.region.AwsRegion
 import software.aws.toolkits.core.utils.tryOrNull
 import software.aws.toolkits.jetbrains.core.AwsResourceCache
+import software.aws.toolkits.jetbrains.core.credentials.activeCredentialProvider
+import software.aws.toolkits.jetbrains.core.credentials.activeRegion
+import software.aws.toolkits.jetbrains.services.redshift.auth.AwsAuth
 import software.aws.toolkits.jetbrains.services.sts.StsResources
+import software.aws.toolkits.jetbrains.ui.CREDENTIAL_ID_PROPERTY
+import software.aws.toolkits.jetbrains.ui.REGION_ID_PROPERTY
 
 private val REDSHIFT_REGION_REGEX = """.*\..*\.(.+).redshift\.""".toRegex()
 private val REDSHIFT_IDENTIFIER_REGEX = """.*//(.+)\..*\..*.redshift\..""".toRegex()
 
 fun extractRegionFromUrl(url: String?): String? = url?.let { REDSHIFT_REGION_REGEX.find(url)?.groupValues?.get(1) }
 fun extractClusterIdFromUrl(url: String?): String? = url?.let { REDSHIFT_IDENTIFIER_REGEX.find(url)?.groupValues?.get(1) }
 
+fun DataSourceRegistry.createDatasource(project: Project, cluster: Cluster) {
+    builder
+        .withJdbcAdditionalProperty(CREDENTIAL_ID_PROPERTY, project.activeCredentialProvider().id)
+        .withJdbcAdditionalProperty(REGION_ID_PROPERTY, project.activeRegion().id)
+        .withUrl(cluster.clusterIdentifier())
+        .withUser(cluster.masterUsername())
+        .withUrl("jdbc:redshift://${cluster.endpoint().address()}:${cluster.endpoint().port()}/${cluster.dbName()}")
+        .commit()
+    // TODO FIX_WHEN_MIN_IS_202 set auth provider ID in builder
+    newDataSources.firstOrNull()?.let {
```

Comment:
what is this magic?

Reply:
So we can't set the authproviderid in the builder (it isn't available). Commit adds it to newDataSources (which is a distinct list from the ones you already have), so we can then pull it back out to set the correct credential ID. I could see if there is a less janky way to do it, but I don't think there is